Rationalize incremental materialization

jtcohen6 opened the pull request:
resolves #140 
resolves #133

### Description

This PR implements, with alterations, several of the recommendations from #140 (much fuller description there) to rationalize the `'incremental'` materialization on `dbt-spark`.

In particular:
- New default strategy, `append`, that provides better consistency with other adapters' default behavior. This strategy works across all file formats, connection types, Spark platforms.
- Raise better errors to enforce which strategies are actually supported when

### Notes

- [x] ~Fix failing `test_dbt_incremental`: I think this PR represents a genuine change in behavior!~ It brings it in line with the default adapter tests, which can only be a step in the right direction :)

- ~I'm not sure about how to implement recommendation (5), which proposed that we should set `file_format: delta` + `incremental_strategy: merge` as defaults if `method == 'odbc'` (replacing `parquet` + `insert_overwrite`, respectively). A bit more on this below:~

_leaving these notes below, even though they're not material to the changes in this PR_

We could add this logic just within the Jinja macros (e.g. within [`file_format_clause`](https://github.com/fishtown-analytics/dbt-spark/blob/2f7c2dcb419b66d463bb83212287bd29280d6964/dbt/include/spark/macros/adapters.sql#L1), [`dbt_spark_validate_get_file_format`](https://github.com/fishtown-analytics/dbt-spark/blob/2f7c2dcb419b66d463bb83212287bd29280d6964/dbt/include/spark/macros/materializations/incremental.sql#L11), [`dbt_spark_validate_get_incremental_strategy`](https://github.com/fishtown-analytics/dbt-spark/blob/2f7c2dcb419b66d463bb83212287bd29280d6964/dbt/include/spark/macros/materializations/incremental.sql#L27)), but then it wouldn't be consistent with the values in `manifest.json`, since the file format default is set in python:

https://github.com/fishtown-analytics/dbt-spark/blob/2f7c2dcb419b66d463bb83212287bd29280d6964/dbt/adapters/spark/impl.py#L32-L33

@kwigley Any chance you could advise on if it would be possible to set that config default dynamically, based on the connection method? Or if that's a terrible idea? :)

Even if we can't get that last piece working, I still think we should move forward with 4/5, and advise Databricks users to simply set in `dbt_project.yml`:
```yml
models:
  +file_format: delta
  +incremental_strategy: merge
```
This is what we advise today, anyway.

### Checklist
 - [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
 - [x] I have run this code in development and it appears to resolve the stated issue
 - [x] This PR includes tests, or tests are not required/relevant for this PR
 - [x] I have updated the `CHANGELOG.md` and added information about my change to the "dbt next" section.

kwigley commented:
> Any chance you could advise on if it would be possible to set that config default dynamically

@jerco unfortunately, I don't think it is possible with the current impl. The adapter config does not know anything about the connection when it is created :(

jtcohen6 commented:
Thanks for looking into it @kwigley!

I suppose this _would_ be a compelling thing that we could do, if we split off the Databricks pieces of `dbt-spark` into a new adapter plugin, `dbt-databricks`, that inherited from the current plugin, added certain functionality, and overrode certain defaults.

Fokko commented on dbt/include/spark/macros/materializations/incremental.sql:
```diff
@@ -85,31 +116,21 @@
-  {% endif %}
-
-  {% call statement() %}
-    set spark.sql.hive.convertMetastoreParquet = false
```
Good to see this good :)

Fokko commented on dbt/include/spark/macros/materializations/incremental.sql:
```diff
@@ -24,59 +35,79 @@
+  {%- elif strategy == 'merge' -%}
+  {#-- merge all columns with databricks delta - schema changes are handled for us #}
     {{ get_merge_sql(target, source, unique_key, dest_columns=none, predicates=none) }}
   {%- endif -%}
```
Maybe raise an error if it doesn't match any of the cases?

jtcohen6 replied:
Good call. It's already raised earlier on:
https://github.com/fishtown-analytics/dbt-spark/blob/c8e3770e077e8c54026156b14e61133ef59fa7ff/dbt/include/spark/macros/materializations/incremental.sql#L65-L66

Just the same, I'll add another explicit exception here, just in case users override some macros but not others.

Fokko commented on dbt/include/spark/macros/materializations/incremental.sql:
```diff
@@ -24,59 +35,79 @@
+    {% if unique_key %}
+        on DBT_INTERNAL_SOURCE.{{ unique_key }} = DBT_INTERNAL_DEST.{{ unique_key }}
+    {% else %}
+        on false
```
This feels awkward, why not just use an `INSERT INTO` statement?

jtcohen6 replied:
This is in line with default dbt behavior [here](https://github.com/fishtown-analytics/dbt/blob/1508564e10175a7562652f93f08aedb5a2845474/core/dbt/include/global_project/macros/materializations/common/merge.sql#L23-L30).

As I see it, the only reason we implement `spark__get_merge_sql` is to benefit from the improved wildcard `update *`/`insert *`.

I agree it feels a bit silly—this is identical to `append` strategy / `INSERT INTO` statement. The alternative is to keep requiring a `unique_key` with `merge`. In any case, I'd rather err closer to the side of default behavior.

Fokko reviewed (commented):
Thanks for unraveling this @jtcohen6. Looking good. I think the most important thing is that we make sure that we test the edge cases. Would love to dive into this a bit deeper, but I'm currently a bit swamped in work :)

jtcohen6 commented:
Thanks for the review @Fokko! I just reorganized these macro files a bit as well.

> I think the most important thing is that we make sure that we test the edge cases.

I completely agree. I have a "suite" of "test cases" in a local project that I used to work through all the edge cases:

| CASE | Apache Spark | Databricks Cluster | Databricks Endpoint |
|-|-|-|-|
| bad_file_format | ❌ | ❌ | ❌ |
| bad_strategy | ❌ | ❌ | ❌ |
| bad_merge_not_delta | ❌ | ❌ | ❌ |
| bad_insert_overwrite_delta | ❌ | ❌ | ❌ |
| default_append | ✅ | ✅ | ❔ |
| insert_overwrite_no_partitions | ✅ | ✅ | ❌ |
| insert_overwrite_partitions | ✅ | ✅ | ❌ |
| delta_append | ❔ | ✅ | ❔ |
| delta_merge_no_key | ❔ | ✅ | ❔ |
| delta_merge_unique_key | ❔ | ✅ | ❔ |

Keys:
- ✅ Works
- ❌ Raises explicit, clear compilation error
- ❔ dbt is unopinionated: no explicit error, may work or may return database error (e.g. `using delta` in Apache Spark, forthcoming support for `create temp view` via endpoint)

I'm struggling a bit with how to implement this in a more-automated fashion:
- It's slightly more involved than a unit test. I tried extending [`unit/test_macros.py`](https://github.com/fishtown-analytics/dbt-spark/blob/master/test/unit/test_macros.py) without success, macros are a bit more interlocking and require dbt-specific Jinja features like `return`
- It's slightly more granular that our suite of adapter integration tests. Ideally we'd have a way to iterate through a sequence, changing one config each time, and catch exceptions where we hope for them.

FWIW We don't have automated tests today for incremental strategies, either, so I think this still constitutes a step in the right direction.

jtcohen6 commented:
As discussed with @kwigley yesterday, the ideal type of test to write for this kind of functionality is a "true" integration test, like the ones we have in Core ([`dbt/test/integration`](https://github.com/fishtown-analytics/dbt/tree/dev/kiyoshi-kuromiya/test/integration)).

To that end, it would be great to include and expose [`base.py`](https://github.com/fishtown-analytics/dbt/blob/dev/kiyoshi-kuromiya/test/integration/base.py) as an available module in `pytest-dbt-adapter`. It's relatively stable boilerplate code that provides the foundation for writing integration or functional tests on any adapter. Breaking it out into its own module is also a prerequisite to someday achieving true modularity for the `dbt-core/-postgres/-redshift/-snowflake/-bigquery` tests.

jtcohen6 commented:
@kwigley I gave a go at writing custom integration tests for the changes in this PR, using the work in https://github.com/fishtown-analytics/dbt-adapter-tests/pull/13. I'd love to get your help here—perhaps a bit of pair-programming? :)

Once we can manage to get these tests running, I'll feel really good about merging and releasing the additions to the adapter testing suite, even if it's missing some of the bells and whistles. (E.g. I've decided that `use_profile` leveraging the `.dbtspec` target, while a very cool idea, is nice but not strictly necessary for this first version—it's the same environment variables, at the end of the day.)

kwigley commented:
@jtcohen6 can you take a peek at why `test_insert_overwrite_apache_spark` is failing?

jtcohen6 commented:
Ugh. Here's where we're at: ODBC connections to Databricks clusters aren't respecting the `set spark.sql.sources.partitionOverwriteMode = DYNAMIC` parameter. I have no idea why. Does `pyodbc` generate a new connection for each executed SQL statement, with new default parameters each time?

(Those `set` statements aren't supported _at all_ by the new SQL Analytics endpoints. This issue is exclusive to `method: odbc` + `cluster`.)

One thing we could do: Tell folks to set `spark.sql.sources.partitionOverwriteMode = DYNAMIC` within their Databricks clusters ([docs](https://docs.databricks.com/clusters/configure.html#spark-config)). I'll try this in a few minutes. If it works, it may be a better idea anyway, rather than having the incremental materialization try to `set` this parameter each time.

kwigley commented on test/custom/incremental_strategies/test_incremental_strategies.py:
```diff
@@ -0,0 +1,102 @@
+        self.assertTablesEqual(
+            "insert_overwrite_no_partitions", "expected_overwrite")
+        self.assertTablesEqual(
+            "insert_overwrite_partitions", "expected_upsert")
```
🤦

jtcohen6 replied:
You're good!!! For a very dumb reason, this _worked_ in Databricks when it really shouldn't have. I fixed via:
<img width="641" alt="Screen Shot 2021-02-19 at 2 50 36 PM" src="https://user-images.githubusercontent.com/13897643/108512661-d6077680-72c1-11eb-8160-3f08a246a977.png">

kwigley reviewed (approved):
thanks for working through the tests with me! I'm happy with shipping this with the intention of spending more time with adapter integration tests.
